### DISCUSSION TITLE
Redirect loop at /undefined

### DISCUSSION THREAD

**Discussion by Keeperr** [outline/outline#1324] [General]
After signing in with Slack, I enter a redirect loop:

https://example.com/undefined?done=159346533644
https://example.com/undefined?done=1593465301265

The `URL` variable in `.env` is set to `https://example.com`.
The redirect URL is set to: `https://example.com/auth/slack.callback`

I have the following setup:

- Ubuntu 20.04 LTS
- Apache/2.4.41
- psql (PostgreSQL) 12.2
- minio version RELEASE.2020-06-22T03-12-50Z
- Redis server v=5.0.7 sha=00000000:0 malloc=jemalloc-5.2.1 bits=64 build=636cde3b5c7a3923
- NodeJS v12.6.0

Using v0.43.3.

**Comment by tommoor** [comment]
Is `SUBDOMAINS_ENABLED` env set to false or missing? (It should be)

**Reply by Keeperr** [reply]
It is.

**Comment by tommoor** [comment]
https://github.com/outline/outline/blob/f43deb7940e01eeb7c4f3ec11c6aa505e39bd747/app/stores/AuthStore.js#L183

This is where the URL is constructed, `BASE_URL` must be undefined – which means your environment variables were not available/set when the webpack build was ran. The env variables must be set at build time.

Hopefully this helps, there are a couple of open tickets to try and remove env variables from the bundled javascript which would make this part of setup simpler.

**Reply by Keeperr** [reply]
Yes, this was the issue. However, now after a login through Slack I end up at the same page (prompting me to log in again) and the URL ends with /?done=1593499582211.

`/home` redirect to `?done=1593499582211` as well, though there is no longer a redirect loop.

**Comment by tommoor** [comment]
Has the user and team been saved successfully in the database?

Would indicate some sort of issue setting the auth cookie, perhaps if
you’re running on a subdomain? Look where the cookie is going, if anywhere.

**Reply by Keeperr** [reply]
No, it's not on a subdomain. My user and team are in the database.

Without `ProxyPassReverseCookiePath / http://example.com:3000` in the Apache configuration I get a state mismatch as the cookies are most probably not being passed.
![state_mismatch](https://i.imgur.com/8ApVzZ7.png)

With `ProxyPassReverseCookiePath / http://example.com:3000` I've got the same cookies set for my domain (`example.com`).

**Reply by tommoor** [reply]
In the developer console, go to:

- Network, and check "preserve log"
- Sign in again
- Take a look at the network requests that were made, did they fail with authentication? Perhaps some clues there.

**Reply by Keeperr** [reply]
Unfortunately, not.

I can attach debug logs for all categories if that would help.
I don't have `AUTH_STORE` in local storage either. Same goes for the `accessToken` cookie - it's not set.

I've given Slack bot access to all scopes.

**Reply by tommoor** [reply]
The authentication is passed to the frontend as a cookie, clearly for some reason this isn't happening – you should be able to see the cookie in the redirect response from `/auth/slack.callback`. My guess is that [the domain here is incorrect](https://github.com/outline/outline/blob/f43deb7940e01eeb7c4f3ec11c6aa505e39bd747/server/middlewares/authentication.js#L92) because you're behind a reverse proxy. 

Make sure that `X-Forwarded-Host` header is set to the correct external domain name so that the app knows where it's hosted.

**Reply by Keeperr** [reply]
I have `X-Forwarded-Host` as well as other reverse proxy-related directives. My current vhost configuration looks like so:

```
<VirtualHost *:443>

        ProxyPreserveHost On
        ProxyRequests off
        SSLProxyEngine on

        ProxyPass / http://example.com:3000/
        ProxyPassReverse / http://example.com:3000/

        ServerName example.com
        #ServerAlias www.example.com

        ServerAdmin me@example.com
        DocumentRoot /var/www/html/outline

        ProxyPassReverseCookiePath / http://example.com:3000

        #RequestHeader set X-Forwarded-Port "3000"
        RequestHeader set X-Forwarded-Proto "https"
        RequestHeader set X-Forwarded-Host "https://example.com"

        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined

        SSLEngine on
        SSLCertificateFile /etc/letsencrypt/live/example.com/fullchain.pem
        SSLCertificateKeyFile /etc/letsencrypt/live/example.com/privkey.pem
        Include /etc/letsencrypt/options-ssl-apache.conf
</VirtualHost>
```

**Reply by Keeperr** [reply]
Fixed it. My current virtual host configuration looks like so for anyone stumbling upon the same problem.

```
<VirtualHost *:443>

        ProxyPreserveHost On
        ProxyRequests off
        SSLProxyEngine on

        ProxyPass / http://example.com:3000/
        ProxyPassReverse / http://example.com:3000/

        ServerName example.com
        ServerAdmin test@example.com
        DocumentRoot /var/www/html/outline

        RequestHeader set X-Forwarded-Proto "https"

        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined

        SSLEngine on

        SSLCertificateFile /etc/letsencrypt/live/example.com/fullchain.pem
        SSLCertificateKeyFile /etc/letsencrypt/live/example.com/privkey.pem
        Include /etc/letsencrypt/options-ssl-apache.conf

</VirtualHost>
```

**Reply by tommoor** [reply]
Glad you got there, and thanks for sharing the final configuration – what was the key in this case?
